blueocean-events/src/main/java/io/jenkins/blueocean/events/PipelineEventListener.java

```diff
@@ -90,45 +89,14 @@ public void onNewHead(FlowNode flowNode) {
         }
     }
 
-    private List<String> getBranch(FlowNode flowNode) {
-        List<String> branch = new ArrayList<>();
-        FlowNode parentBlock = getParentBlock(flowNode);
-
-        while (parentBlock != null) {
-            branch.add(0, parentBlock.getId());
-            parentBlock = getParentBlock(parentBlock);
-        }
-
-        return branch;
-    }
-
-    private FlowNode getParentBlock(FlowNode flowNode) {
-        List<FlowNode> parents = flowNode.getParents();
-
-        for (FlowNode parent : parents) {
-            if (parent instanceof StepStartNode) {
-                if (parent.getAction(BodyInvocationAction.class) != null) {
-                    return parent;
-                }
-            }
-        }
-
-        for (FlowNode parent : parents) {
-            if (parent instanceof StepEndNode) {
-                continue;
-            }
-            FlowNode grandparent = getParentBlock(parent);
-            if (grandparent != null) {
-                return grandparent;
-            }
-        }
-
-        return null;
+    /* package: so that we can unit test it */ List<String> getBranch(FlowNode flowNode) {
+        return Lists.reverse(flowNode.getEnclosingBlocks().stream()
+                .map(FlowNode::getId).collect(Collectors.toList()));
     }
 
-    private String toPath(List<String> branch) {
+    private String toPath(Collection<String> branch) {
         StringBuilder builder = new StringBuilder();
-        for (String leaf : branch) {
+        for(String leaf: branch){
```

Comment:
🐜  Can this path creation with just `branch.stream().collect(Collectors.joining("/"))` if you're feeling all Streams-y :)

Reply:
Right, there are many places in the code we need to change to adopt java 8, keeping the scope limited to just this fix, so be making such changes some other time.